pkg/provider/cloud/openstack/provider.go

```diff
@@ -195,233 +199,377 @@ func validateExistingSubnetOverlap(networkID string, netClient *gophercloud.Serv
-			finalizers = append(finalizers, RouterSubnetLinkCleanupFinalizer)
+		if existingNetwork != nil {
+			return cluster, nil
 		}
 	}
-	if ipv6Network {
-		// If subnet has to be created, add associated finalizer.
-		if cluster.Spec.Cloud.Openstack.IPv6SubnetID == "" {
-			finalizers = append(finalizers, IPv6SubnetCleanupFinalizer)
-		}
-		// If subnet or router has to be created, subnet needs to be attached to the router.
-		if cluster.Spec.Cloud.Openstack.IPv6SubnetID == "" || (cluster.Spec.Cloud.Openstack.RouterID == "" && routerID == "") {
-			finalizers = append(finalizers, RouterIPv6SubnetLinkCleanupFinalizer)
+
+	// If NetworkName not specified, Create network with name kubernetes-clusterid
+	if networkName == "" {
+		networkName = cluster.Name
+	}
+
+	newNetwork, err := createKubermaticNetwork(netClient, networkName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create the kubermatic network: %w", err)
+	}
+	cluster, err = update(ctx, cluster.Name, func(cluster *kubermaticv1.Cluster) {
+		kubernetes.AddFinalizer(cluster, NetworkCleanupFinalizer)
+		cluster.Spec.Cloud.Openstack.Network = newNetwork.Name
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to update network cluster: %w", err)
+	}
+	return cluster, nil
+}
+
+func reconcileSecurityGroup(ctx context.Context, netClient *gophercloud.ServiceClient, cluster *kubermaticv1.Cluster, update provider.ClusterUpdater) (*kubermaticv1.Cluster, error) {
+	securityGroup := cluster.Spec.Cloud.Openstack.SecurityGroups
+
+	// if we already have an ID on the cluster, check if that group still exists
+	if securityGroup != "" {
+		err := validateSecurityGroupsExist(netClient, []string{securityGroup})
+		if err != nil {
+			if !isNotFoundErr(err) {
+				return cluster, fmt.Errorf("failed to get security groups: %w", err)
+			}
+		} else {
+			return cluster, nil
 		}
 	}
+	if securityGroup == "" {
+		securityGroup = resourceNamePrefix + cluster.Name
+	}
+
+	lowPort, highPort := resources.NewTemplateDataBuilder().
+		WithNodePortRange(cluster.Spec.ComponentsOverride.Apiserver.NodePortRange).
+		WithCluster(cluster).
+		Build().
+		NodePorts()
+
+	ipv4Network := cluster.IsIPv4Only() || cluster.IsDualStack()
+	ipv6Network := cluster.IsIPv6Only() || cluster.IsDualStack()
+
+	req := createKubermaticSecurityGroupRequest{
+		secGroupName: securityGroup,
+		ipv4Rules:    ipv4Network,
+		ipv6Rules:    ipv6Network,
+		lowPort:      lowPort,
+		highPort:     highPort,
+	}
+
+	req.nodePortsCIDRs = resources.GetNodePortsAllowedIPRanges(cluster, cluster.Spec.Cloud.Openstack.NodePortsAllowedIPRanges, cluster.Spec.Cloud.Openstack.NodePortsAllowedIPRange)
 
-	cluster, err = ensureFinalizers(ctx, cluster, finalizers, update)
+	secGroupName, err := createKubermaticSecurityGroup(netClient, req)
 	if err != nil {
-		return nil, fmt.Errorf("failed to ensure finalizers: %w", err)
+		return nil, fmt.Errorf("failed to create the kubermatic security group: %w", err)
```

Comment:
Same question regarding "kubermatic SG" here. Should this be "user cluster SG" instead?